working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+- the `validate_book` function definition
+- usage in `read_one_book`
+- usage in `update_book`
+- usage in `delete_book`
+
+Peeking at our test suite in `test_routes.py`, we do have tests for `read_one_book`, but we need to write tests for the remaining untested dependencies: `validate_book`, `update_book`, and `delete_book`. 
+
+### Updating Our Test Suite
+
+#### Testing `validate_book`
+
+We'll start with testing the `validate_book` function. Before we begin, we're going to open several files to ensure Copilot has plenty of context around our code and test set up. Take a moment to ensure the following files are open in VS Code:
+- `book.py`
+- `routes.py`
+- `conftest.py`
+- `test_models.py`
+- `test_routes.py`
+
+We're going to start from `routes.py` and use our cursor to highlight the entire `validate_book` function. Next, bring up the inline Copilot chat and invoke the `/tests` shortcut. When Copilot is done thinking, we may see a slightly different UI than in previous scenarios:
+
+![A VS Code window with the routes.py file up and validate_book function selected showing a Refactor Preview tab at the bottom of the screen](assets/copilot-in-projects/routes-test-validate-book.png)  
+*Fig. Copilot's `Refactor Preview` tab showing at the bottom of VS Code*
+
+This time instead of showing changes to the current file in a temporary window, the bottom panel has opened to a `Refactor Preview` tab that has a single entry on it. If we click the entry, it will take us to a temporary file showing the changes Copilot is suggesting that we make to the `test_routes.py` file. 
+
+![A VS Code window with a temporary file showing that contains the original contents of test_routes.py along with suggested changes to test the `validate_book` function](assets/copilot-in-projects/test-routes-validate-book-refactor-preview.png)  
+*Fig. Temporary file in VS Code showing the test suggestions from Copilot*
+
+Here we can review the tests cases Copilot generated and see if there are changes or further tests we need. We have "Apply" and "Discard" controls at the bottom of the screen in the `Refactor Preview` tab to either accept or reject the changes.
+
+If we examine the code Copilot created, the scenarios identified are great – we have tests for the nominal case and a couple edge cases of invalid or non-existent book ids. However, there are a number of issues:
+- imports were added in the middle of the file, one of which is tacked on to the last line of an existing test
+- `test_validate_book_with_valid_id` doesn't use our `two_saved_books` test fixture to ensure that there are books in the database before the test runs 
+- `test_validate_book_with_valid_id` thinks the result of `validate_book(book_id)` will be an integer rather than a `Book` instance  
+- `test_validate_book_with_invalid_id` and `test_validate_book_with_nonexistent_id` are trying to read the error value from outside the `with pytest raises` block where it's out of scope
+
```

Comment:
- maybe me being paranoid, but since I got some different results this seemed like a good place to drop a reminder that students could get different results
```suggestion

### !callout-info

## Be Sure to Review Your Own Output

Remember, the tests generated during the Copilot run in this lesson may not match what you see when trying on your own. Always closely review any suggestions from Copilot or any other AI tool to ensure they meet your needs and are correct for your project.

### !end-callout

```